### PR TITLE
Initial CLI component tests: logs, whoami, browse

### PR DESCRIPTION
**Pull request by arielshaqed**
These test against a fake Lycan, allowing focus on fine-grained behaviour.

### CODE REVIEW

File: cli/package.json [hunk 1]
```diff
@@ -70,6 +70,7 @@
     "validator": "^11.1.0"
   },
   "devDependencies": {
+    "@binaris/spice-koa-server": "~0.0.2",
```
Comment (bergundy) [inline cli/package.json]: Can we have carets please?

File: cli/package.json [hunk 2]
```diff
@@ -87,6 +88,7 @@
     "globby": "^10.0.1",
     "ramda": "~0.26.1",
     "specshell": "^0.0.1-rc.2",
+    "testdouble": "~3.12.4",
```
Comment (bergundy) [inline cli/package.json]: Can we have carets please?
See also `ramda` above..

File: cli/src/test/fake_lycan.ts [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+  - directory: ${t.context.projectDir}
+    applicationId: fluffy-samaritan
+    defaultEnv: default
+`);
+    await mkdir(t.context.projectDir);
+    // CLI only needs package.json to mark a project root.
+    await writeFile(path.join(t.context.projectDir, 'package.json'), '');
+  });
+
+  test.after(async (t) => {
```
Comment (bergundy) [inline cli/src/test/fake_lycan.ts]: `.always` please

File: cli/src/test/logs.spec.ts [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+import * as td from 'testdouble';
+import { Context, addFake } from './fake_lycan';
+import * as R from 'ramda';
+import { success } from 'specshell';
+
+const test = anyTest as TestInterface<Context>;
+
+addFake(test);
+
+const process = R.evolve({ out: (x: Buffer) => x.toString(), err: (x: Buffer) => x.toString() });
```
Comment (bergundy) [inline cli/src/test/logs.spec.ts]: No need for this anymore

File: cli/src/test/browse.spec.ts [hunk 5]
```diff
@@ -0,0 +1,27 @@
+import anyTest, { TestInterface } from 'ava';
+import * as td from 'testdouble';
+import { Context, addFake } from './fake_lycan';
+import * as R from 'ramda';
+
+const test = anyTest as TestInterface<Context>;
+
+addFake(test);
+
+const process = R.evolve({ out: (x: Buffer) => x.toString(), err: (x: Buffer) => x.toString() });
```
Comment (bergundy) [inline cli/src/test/browse.spec.ts]: No need for this anymore

File: cli/src/test/logs.spec.ts [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+
+const fakeLogs = [
+  { source: 'foo', msg: 'main engine fire\n', isErr: false, time: new Date('1977-09-05T12:55:55') },
+  { source: 'bar', msg: 'liftoff\n', isErr: true, time: new Date('1977-09-05T12:56:00'),
+    invocation: { durUs: 10000, reqid: 'launch-voyager-1' } },
+  { source: 'space', msg: 'Amalthea flyby\n', isErr: false, time: new Date('1979-03-05T06:54:00') },
+  { source: 'space', msg: 'Jupiter closest approach\n', isErr: false, time: new Date('1979-03-05T12:05:26') },
+];
+
+// test.skip('no logs for incorrect app id', () => 1);
```
Comment (bergundy) [inline cli/src/test/logs.spec.ts]: Leftover?

File: cli/src/test/fake_lycan.ts [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+  lycanUrl: string;
+  lycanFake: td.TestDouble<LycanHandler>;
+  lycanServer: LycanServer;
+  server: Server;
+}
+
+// Adds before and after hooks to help test with a shell against a
+// fake Lycan server.
+export function addFake<C extends Context>(test: TestInterface<C>) {
+  test.before(async (t) => {
```
Comment (vogre) [inline cli/src/test/fake_lycan.ts]: Possible to use sync functions in `before` and `after` for simplicity.

File: cli/src/test/fake_lycan.ts [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+    t.context.shell = new Shell(undefined, {
+      env: {
+        ...processEnv,
+        SHIFTJS_CONFIG: t.context.configPath,
+        SHIFTJS_API_ENDPOINT: t.context.lycanUrl,
+      }});
+    t.assert(success(await t.context.shell.run(`cd ${t.context.projectDir}`)));
+  });
+
+  test.afterEach((t) => { t.context.server.close(); });
```
[the remark this answers is not in the document]
Comment (arielshaqed) [reply cli/src/test/fake_lycan.ts]: Done.  If there's a server.

File: cli/src/test/fake_lycan.ts [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+  lycanServer: LycanServer;
+  server: Server;
+}
+
+// Adds before and after hooks to help test with a shell against a
+// fake Lycan server.
+export function addFake<C extends Context>(test: TestInterface<C>) {
+  test.before(async (t) => {
+    // Quoted in case dirname includes spaces etc.
+    t.context.run = `'${path.resolve(__dirname, '../..', 'bin/run')}'`;
```
Comment (vogre) [inline cli/src/test/fake_lycan.ts]: Needs to be shell-escaped for paths with spaces.

File: cli/src/test/logs.spec.ts (hunk 4 above)
Comment (arielshaqed) [reply cli/src/test/logs.spec.ts]: Done.

File: cli/src/test/logs.spec.ts (hunk 6 above)
Comment (arielshaqed) [reply cli/src/test/logs.spec.ts]: Removed

File: cli/src/test/fake_lycan.ts (hunk 9 above)
Comment (arielshaqed) [reply cli/src/test/fake_lycan.ts]: Good idea!  It's quoted, but that's obviously not nearly as good.

**Pull request merged**